--- a/config.def.h
+++ a/config.def.h
@@ -82,6 +82,14 @@
         } \
 }
 
+/* YTPLAY(URI) */
+#define YTPLAY(u) {\
+        .v = (const char *[]){ "/bin/sh", "-c", \
+             "youtube-dl -o - \"$0\" | mpv --really-quiet -", u, NULL \
+        } \
+}
+
+
 /* styles */
 /*
  * The iteration will stop at the first match, beginning at the beginning of
@@ -105,6 +113,7 @@
 	{ MODKEY,                GDK_KEY_f,      spawn,      SETPROP("_SURF_FIND", "_SURF_FIND") },
 	{ MODKEY,                GDK_KEY_slash,  spawn,      SETPROP("_SURF_FIND", "_SURF_FIND") },
 
+        { MODKEY,                GDK_KEY_w,      ytplayexternal, { 0 } },
 	{ 0,                     GDK_KEY_Escape, stop,       { 0 } },
 	{ MODKEY,                GDK_KEY_c,      stop,       { 0 } },

--- a/surf.c
+++ a/surf.c
@@ -217,6 +217,7 @@
 static void togglecookiepolicy(Client *c, const Arg *a);
 static void toggleinspector(Client *c, const Arg *a);
 static void find(Client *c, const Arg *a);
+static void ytplayexternal(Client *c, const Arg *a);

 /* Buttons */
 static void clicknavigate(Client *c, const Arg *a, WebKitHitTestResult *h);
@@ -1645,6 +1646,15 @@
        Arg arg;

        arg = (Arg)VIDEOPLAY(webkit_hit_test_result_get_media_uri(h));
+       spawn(c, &arg);
+}
+
+void
+ytplayexternal(Client *c, const Arg *a)
+{
+       Arg arg;
+
+       arg = (Arg)YTPLAY(geturi(c));
        spawn(c, &arg);
 }
