diff --git a/meson.build b/meson.build
index 3e61584..b5dd042 100644
--- a/meson.build
+++ b/meson.build
@@ -103,9 +103,7 @@ if get_option('with_wayland') != 'no'
   wayland_dep = dependency('wayland-client', version : '>= 1.11.0',
 			   required : get_option('with_wayland') == 'yes')
   if wayland_dep.found()
-    wl_prefix = wayland_dep.get_variable(pkgconfig: 'prefix')
-    wl_scanner = find_program('wayland-scanner',
-			      wl_prefix + '/bin/wayland-scanner')
+    wl_scanner = find_program('wayland-scanner', '/usr/bin/wayland-scanner')
   endif
   WITH_WAYLAND = wayland_dep.found()
 endif
