diff --git a/layout.mobintl.h b/layout.mobintl.h
index 6e02c78..7d763f7 100644
--- a/layout.mobintl.h
+++ b/layout.mobintl.h
@@ -986,7 +986,7 @@ static struct key keys_landscape[] = {
   {"Tab", "Tab", 0.75, Code, KEY_TAB, .scheme = 1},
   {"", "", 0.0, EndRow},
 
-  {"Cmp", "Cmp", 1.5, Compose, .scheme = 1},
+  {"Sup", "Sup", 1.5, Mod, Super, .scheme = 1},
   {"a", "A", 1.0, Code, KEY_A, &layouts[ComposeA]},
   {"s", "S", 1.0, Code, KEY_S, &layouts[ComposeS]},
   {"d", "D", 1.0, Code, KEY_D, &layouts[ComposeD]},
