--- a/vendor/openssl-src/src/lib.rs
+++ a/vendor/openssl-src/src/lib.rs
@@ -256,7 +256,7 @@
             "powerpc-unknown-linux-gnu" => "linux-ppc",
             "powerpc64-unknown-freebsd" => "BSD-generic64",
             "powerpc64-unknown-linux-gnu" => "linux-ppc64",
-            "powerpc64-unknown-linux-musl" => "linux-ppc64",
+            "powerpc64-unknown-linux-musl" => "linux-generic64",
             "powerpc64le-unknown-freebsd" => "BSD-generic64",
             "powerpc64le-unknown-linux-gnu" => "linux-ppc64le",
             "powerpc64le-unknown-linux-musl" => "linux-ppc64le",
