--- a/configure.orig
+++ a/configure
@@ -5803,30 +5803,6 @@
 
 
 
-# Clang sanitizers wrap reallocarray even if it isn't available on the target
-# system. When compiled it always returns NULL and crashes the program. To
-# detect this we need a more complicated test.
-{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for working reallocarray" >&5
-printf %s "checking for working reallocarray... " >&6; }
-if test "$cross_compiling" = yes
-then :
-  { { printf "%s\n" "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
-printf "%s\n" "$as_me: error: in \`$ac_pwd':" >&2;}
-as_fn_error $? "cannot run test program while cross compiling
-See \`config.log' for more details" "$LINENO" 5; }
-else $as_nop
-  cat confdefs.h - <<_ACEOF >conftest.$ac_ext
-/* end confdefs.h.  */
-#include <stdlib.h>
-int
-main (void)
-{
-return (reallocarray(NULL, 1, 1) == NULL);
-
-  ;
-  return 0;
-}
-_ACEOF
 if ac_fn_c_try_run "$LINENO"
 then :
   { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: yes" >&5
