diff --git a/protocol/meson.build b/protocol/meson.build
index 4d56bbf..bc93f0a 100644
--- a/protocol/meson.build
+++ b/protocol/meson.build
@@ -5,11 +5,7 @@ wayland_protos = dependency('wayland-protocols',
 )
 wl_protocol_dir = wayland_protos.get_variable('pkgdatadir')
 
-wayland_scanner_dep = dependency('wayland-scanner', native: true)
-wayland_scanner = find_program(
-	wayland_scanner_dep.get_variable('wayland_scanner'),
-	native: true,
-)
+wayland_scanner = find_program('wayland-scanner')
 
 protocols = {
 	# Stable upstream protocols
