--- configure.ac.orig
+++ configure.ac
@@ -59,14 +59,14 @@
 # Set-up variables to use build machine compiler when cross-compiling
 if test x"$CC_FOR_BUILD" = x; then
 	if test x"$cross_compiling" = xyes; then
-		AC_CHECK_PROGS(CC_FOR_BUILD, gcc cc)
+		AC_CHECK_PROGS(CC_FOR_BUILD, clang cc)
 	else
 		CC_FOR_BUILD="$CC"
 	fi
 fi
 if test x"$cross_compiling" != xyes; then
-       : ${CFLAGS_FOR_BUILD=${CFLAGS}}
-       : ${LDFLAGS_FOR_BUILD=${LDFLAGS}}
+       : ${CFLAGS_FOR_BUILD=""}
+       : ${LDFLAGS_FOR_BUILD=""}
 fi
 AC_SUBST([CC_FOR_BUILD])
 DEFAULT_CPPFLAGS_FOR_BUILD="${CPPFLAGS}"
