--- a/Makefile.orig
+++ a/Makefile
@@ -159,18 +159,18 @@
 # Toolchain
 ################################################################################
 
-HOSTCC			:=	gcc
+HOSTCC			:=	clang
 export HOSTCC
 
-CC			:=	${CROSS_COMPILE}gcc
+CC			:=	clang --target=aarch64-linux-musl
 CPP			:=	${CROSS_COMPILE}cpp
-AS			:=	${CROSS_COMPILE}gcc
+AS			:=	${CROSS_COMPILE}${CC}
 AR			:=	${CROSS_COMPILE}ar
 LINKER			:=	${CROSS_COMPILE}ld
 OC			:=	${CROSS_COMPILE}objcopy
 OD			:=	${CROSS_COMPILE}objdump
 NM			:=	${CROSS_COMPILE}nm
-PP			:=	${CROSS_COMPILE}gcc -E
+PP			:=	${CROSS_COMPILE}${CC} -E
 DTC			:=	dtc
 
 # Use ${LD}.bfd instead if it exists (as absolute path or together with $PATH).
@@ -278,7 +278,7 @@
 	ifeq (${ARCH},aarch64)
 		LTO_CFLAGS	=	-flto
 		# Use gcc as a wrapper for the ld, recommended for LTO
-		LINKER		:=	${CROSS_COMPILE}gcc
+		LINKER		:=	${CROSS_COMPILE}${CC}
 	endif
 endif
 LD			=	$(LINKER)
