--- a/plat/rockchip/rk3399/drivers/m0/Makefile.orig	2021-09-13 23:47:12.000000000 +0100
+++ a/plat/rockchip/rk3399/drivers/m0/Makefile	2021-09-14 00:02:03.810000000 +0100
@@ -36,7 +36,7 @@
 COMMON_FLAGS		:= -g -mcpu=$(ARCH) -mthumb -Wall -O3 -nostdlib -mfloat-abi=soft
 CFLAGS			:= -ffunction-sections -fdata-sections -fomit-frame-pointer -fno-common
 ASFLAGS			:= -Wa,--gdwarf-2
-LDFLAGS			:= -Wl,--gc-sections -Wl,--build-id=none
+LDFLAGS			:= -Wl,--gc-sections -Wl,--build-id=none -Wl,--no-dynamic-linker
 
 # Cross tool
 CC			:= ${M0_CROSS_COMPILE}gcc
