Allow build with clang
--- a/configure.ac	2021-01-10 22:10:11.018000000 +0000
+++ a/configure.ac	2021-01-10 22:10:30.162000000 +0000
@@ -1206,13 +1206,6 @@
 fi
 
 AC_CHECK_TOOLS([READELF], [readelf], [:])
-if test "$cross_compiling" = yes; then
-    case "$READELF" in
-	readelf|:)
-	AC_MSG_ERROR([readelf for the host is required for cross builds])
-	;;
-    esac
-fi
 AC_SUBST(READELF)
 
 
--- a/configure	2021-01-10 22:53:22.469000000 +0000
+++ a/configure	2021-01-10 22:53:33.414000000 +0000
@@ -6205,15 +6205,6 @@
   fi
 fi
 
-if test "$cross_compiling" = yes; then
-    case "$READELF" in
-	readelf|:)
-	as_fn_error $? "readelf for the host is required for cross builds" "$LINENO" 5
-	;;
-    esac
-fi
-
-
 
 case $MACHDEP in
 hp*|HP*)
