--- a/Configurations/10-main.conf.orig
+++ a/Configurations/10-main.conf
@@ -714,6 +714,12 @@
         perlasm_scheme   => "linux32",
         lib_cppflags     => add("-DB_ENDIAN"),
     },
+    "linux-ppcle" => {
+        inherit_from     => [ "linux-latomic" ],
+        asm_arch         => 'ppc32',
+        perlasm_scheme   => "linux32",
+        lib_cppflags     => add("-DL_ENDIAN"),
+    },
     "linux-ppc64" => {
         inherit_from     => [ "linux-generic64" ],
         cflags           => add("-m64"),
